### PR TITLE
[Dist CI][BE] Run each test in its own process for test_distributed_spawn

### PR DESCRIPTION
**Pull request by rohan-varma**
Context: https://github.com/pytorch/pytorch/issues/67061

Use `run_test.py`'s provided flag `"--subprocess"`, passed in like `extra_unittest_args=["--subprocess"]` when running test_distributed_spawn. This will ensure that each test is run separately in its own process. The goal is to more closely simulate how a developer would run a single test when reproducing a CI failure and make reproducibility easier in general.

Also, when a test fails, print out the exact command that was issued so developer knows how to reproduce it.

For example test fails, it will print out something like the following to logs - 

```
Test exited with non-zero exitcode 1. Command to reproduce: BACKEND=gloo WORLD_SIZE=3 /fsx/users/rvarm1/conda/envs/pytorch/bin/python distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_Backend_enum_class
```

running test_distributed_spawn is still the same cmd as before:

`
python test/run_test.py --verbose -i distributed/test_distributed_spawn
`

as seen in [distributed contributing](https://github.com/pytorch/pytorch/blob/master/torch/distributed/CONTRIBUTING.md) guide.

### CODE REVIEW

File: test/run_test.py [hunk 1]
```diff
@@ -596,7 +596,7 @@ def test_distributed(test_module, test_directory, options):
                         test_module, test_directory, options, launcher_cmd=mpiexec
                     )
                 else:
-                    return_code = run_test(test_module, test_directory, options)
+                    return_code = run_test(test_module, test_directory, options, extra_unittest_args=["--subprocess"])
```
Comment (janeyx99) [inline test/run_test.py:599]: 🎉

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -577,8 +577,18 @@ def run_tests(argv=UNITTEST_ARGS):
         failed_tests = []
         for case in test_cases:
             test_case_full_name = case.id().split('.', 1)[1]
-            exitcode = shell([sys.executable] + argv + [test_case_full_name])
+            cmd = [sys.executable] + argv + [test_case_full_name]
+            string_cmd = " ".join(cmd)
+            exitcode = shell(cmd)
             if exitcode != 0:
+                # This is sort of hacky, but add on relevant env variables for distributed tests.
+                if 'TestDistBackendWithSpawn' in test_case_full_name:
```
Comment (cbalioglu) [inline torch/testing/_internal/common_utils.py:585]: Should we stop testing spawn altogether? I think in the internal tests we already removed spawn right?

File: test/run_test.py (hunk 1 above)
Comment (cbalioglu) [reply test/run_test.py:599]: Neat :)

**Review by cbalioglu — APPROVED** [review]
LGTM!

**Review by mruberry — APPROVED** [review]
Cool! Excited to see the effect this has

**Comment by rohan-varma** [comment]
`21:00:35 FAIL [0.433s]: test_nadam (__main__.TestOptim)` is unrelated on ROCM CI.

**Pull request closed without merging**